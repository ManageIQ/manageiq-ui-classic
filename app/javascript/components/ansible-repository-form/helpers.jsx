const GitUrlParse = require('git-url-parse');

export const urlValidation = (values) => {
  const errors = {};
  var error = false;
  try {
    const parsed = values.scm_url ? GitUrlParse(values.scm_url) : {};
    if ((parsed.protocol !== 'http'
          && parsed.protocol !== 'https'
          && parsed.protocol !== 'file'
          && parsed.protocol !== 'ssh')
          || parsed.pathname === '') {
      error = true;
    }
    if (parsed.protocols.length > 2 || (parsed.protocols.length === 1 &&  parsed.protocols[0] !== parsed.protocol)) {
      error = true;
    }
  } catch (err) {
    error = true;
  }
  if (error) {
    errors.scm_url = __('URL must include a protocol (http://, https:// or file://) with path or be a valid SSH path (user@server:path or ssh://user@address:port/path)');
  }
  return errors;
};
