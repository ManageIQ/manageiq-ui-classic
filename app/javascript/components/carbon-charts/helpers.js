export const sampleData = [{
  group: 'Dataset 1',
  key: '1',
  value: 100,
},
{
  group: 'Dataset 1',
  key: '2',
  value: 400,
},
{
  group: 'Dataset 1',
  key: '4',
  value: 500,
},
{
  group: 'Dataset 1',
  key: '3',
  value: 200,
},
{
  group: 'Dataset 1',
  key: '5',
  value: 150,
},
{
  group: 'Dataset 2',
  key: '1',
  value: 280,
},
{
  group: 'Dataset 2',
  key: '2',
  value: 350,
},
{
  group: 'Dataset 2',
  key: '3',
  value: 700,
},
{
  group: 'Dataset 2',
  key: '5',
  value: 600,
},
{
  group: 'Dataset 2',
  key: '5',
  value: 200,
},
{
  group: 'Dataset 3',
  key: '4',
  value: 100,
},
{
  group: 'Dataset 3',
  key: '2',
  value: 500,
},
{
  group: 'Dataset 3',
  key: '1',
  value: 200,
}];

export const pieData = [{
  group: 'Dataset 1',
  key: '1',
  value: 100,
},
{
  group: 'Dataset 2',
  key: '1',
  value: 280,
},
{
  group: 'Dataset 3',
  key: '4',
  value: 100,
},
{
  group: 'Dataset 4',
  key: '2',
  value: 500,
}];

// convert report data to carbon chart data format.
export const getConvertedData = (data) => {
  if (data && data.data.columns && data.data && data.miq && data.miq.category_table && data.miq.name_table) {
    const columnsData = data.data.columns;
    const dataGroups = data.miq.category_table;
    const rowsData = data.miq.name_table;
    const arr = [];
    if (columnsData.length > 0) {
      columnsData.forEach((items) => {
        items.forEach((item, i) => {
          const obj = {};
          if (i !== 0 && rowsData[items[0]]) {
            obj.group = rowsData[items[0]];
            obj.key = dataGroups[i - 1];
            obj.value = item;
            arr.push(obj);
          }
        });
      });
    }
    return arr;
  }
  return [];
};

export const getLineConvertedData = (data) => {
  const columnsData = data.data.columns;
  const dataGroups = data.miq.category_table;
  const rowsData = data.miq.name_table;
  const arr = [];
  columnsData.forEach((items) => {
    items.forEach((item, i) => {
      const obj = {};
      if (i !== 0 && rowsData[items[0]]) {
        obj.group = rowsData[items[0]];
        obj.key = new Date(dataGroups[i - 1]);
        obj.value = item;
        arr.push(obj);
      }
    });
  });
  return arr;
};
