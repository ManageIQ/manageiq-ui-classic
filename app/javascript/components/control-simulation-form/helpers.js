// Helper Functions

var showPassedPolicies = true;
var showFailedPolicies = true;
var showOutOfScopeItems = true;

const convertApi = (apiData, showPassedPolicies, showFailedPolicies, showOutOfScopeSelected) => {
  showPassedPolicies = showPassedPolicies;
  showFailedPolicies = showFailedPolicies;
  showOutOfScopeItems = showOutOfScopeSelected;

  if (apiData === undefined) {
    return '';
  }
  console.log('api Data');
  console.log(apiData);

  const convToString = [{
    key: 'root',
    text: apiData.name,
    icon: 'fa fa-star',
    class: 'no-cursor',
    selectable: false,
    state: {
      expanded: true,
    },
    nodes: findAllVms(apiData.task_results),
  }];
  return convToString;
};

const findAllVms = (allVms) => {
  // TODO: test this, there are cases with more than one vm (By Clusters -> Cluster)
  const holdAllVms = [];

  allVms.forEach((vmData) => {
    const vmObj = {
      key: `xx-${vmData.id}`,
      text: `\u003cstrong\u003eVM:\u003c/strong\u003e ${vmData.name}`,
      icon: 'pficon pficon-virtual-machine',
      class: 'no-cursor',
      selectable: false,
      nodes: findAllProfiles(vmData.profiles, `xx-${vmData.id}`),
      state: {
        expanded: false,
      },
    };
    holdAllVms.push(vmObj);
  });
  return holdAllVms;
};

const findAllProfiles = (allProfiles, upperKey) => {
  const holdAllPolicies = [];

  allProfiles.forEach((profileData) => {
    const profileObj = {
      key: `${upperKey}_xx-${profileData.id}`,
      text: `\u003cstrong\u003eProfile:\u003c/strong\u003e ${profileData.name}`,
      icon: determineIcon(profileData.result),
      class: 'no-cursor',
      selectable: false,
      nodes: findAllPolicies(profileData.policies, `${upperKey}_xx-${profileData.id}`),
      state: {
        expanded: false,
      },
    };
    holdAllPolicies.push(profileObj);
  });
  return holdAllPolicies;
};

const findAllPolicies = (allPolicies, upperKey) => {
  const holdAllPolicies = [];
  allPolicies.forEach((policyData) => {
    // NOTE: you cannot give the tree a {} option it will break
    const nodes = [];
    if (policyData.scope) {
      nodes.push(findAllScope(policyData.scope, `${upperKey}_xx-${policyData.id}`));
    }
    if (policyData.conditions.length !== 0) {
      nodes.push(findAllConditions(policyData.conditions, `${upperKey}_xx-${policyData.id}`));
    }
    if (policyData.actions.length !== 0) {
      nodes.push(findAllActions(policyData.actions, `${upperKey}_xx-${policyData.id}`));
    }
    const policyObj = {
      key: `${upperKey}_xx-${policyData.id}`,
      text: `\u003cstrong\u003ePolicy:\u003c/strong\u003e ${policyData.description}`,
      icon: determineIcon(policyData.result),
      class: 'no-cursor',
      selectable: false,
      nodes,
      state: {
        expanded: false,
      },
    };

    // NOTE: policyData.result can be: deny || allow || N/A
    if (policyData.result === 'deny' && showFailedPolicies === true) {
      holdAllPolicies.push(policyObj);
    } else if (policyData.result === 'allow' && showPassedPolicies === true) {
      holdAllPolicies.push(policyObj);
    } else if (policyData.result === 'N/A' && showOutOfScopeItems === true) {
      holdAllPolicies.push(policyObj);
    }
  });
  return holdAllPolicies;
};

const findAllConditions = (allConditions, upperKey) => {
  let holdAllConditions = {};

  allConditions.forEach((conditionData) => {
    const conditionObj = {
      key: `${upperKey}_xx-${conditionData.id}`,
      text: `\u003cstrong\u003eCondition:\u003c/strong\u003e ${conditionData.description}`,
      icon: determineIcon(conditionData.result),
      class: 'no-cursor',
      selectable: false,
      nodes: [findAllExpressions(conditionData.expression, `${upperKey}_xx-${conditionData.id}`)],
      state: {
        expanded: false,
      },
    };
    holdAllConditions = conditionObj;
  });
  return holdAllConditions;
};

const findAllActions = (allActions, upperKey) => {
  let holdAllActions = {};

  allActions.forEach((actionData) => {
    const actionObj = {
      key: `${upperKey}_xx-${actionData.id}`,
      text: `\u003cstrong\u003eAction:\u003c/strong\u003e ${actionData.description}`,
      icon: determineIcon(actionData.result),
      class: 'no-cursor',
      selectable: false,
      state: {
        expanded: false,
      },
    };
    holdAllActions = actionObj;
  });
  return holdAllActions;
};

const findAllExpressions = (allExpressions, upperKey) => {
  const dynamicallyBuildText = createSentance(allExpressions, '');
  const holdExpression = {
    key: `${upperKey}_xx-`, // NOTE: this doesnt have an id
    text: `\u003cstrong\u003eExpression:\u003c/strong\u003e ${determineTextColor(allExpressions.result, dynamicallyBuildText)}`,
    tooltip: dynamicallyBuildText,
    icon: determineIcon(allExpressions.result),
    class: 'no-cursor',
    selectable: false,
    state: {
      expanded: false,
    },
  };
  return holdExpression;
};

const findAllScope = (allScope, upperKey) => {
  const dynamicallyBuildText = createSentance(allScope, '');
  const holdScope = {
    key: `${upperKey}_xx-`, // NOTE: this doesnt have an id
    text: `\u003cstrong\u003eScope:\u003c/strong\u003e ${determineTextColor(allScope.result, dynamicallyBuildText)}`,
    tooltip: dynamicallyBuildText,
    icon: determineIcon(allScope.result), // TODO: Is scope always red x ?
    class: 'no-cursor',
    selectable: false,
    state: {
      expanded: false,
    },
  };
  return holdScope;
};

export { convertApi };

// ==================================================================== //

const createSentance = (object, runOnSentence) => {
  // TODO: Test more
  let result = runOnSentence;
  // var useAnd = result == "" ? "" : " && ";
  for (const key in object) {
    if (key == 'result') {
      return result;
    } if (object[key].field != undefined) {
      result = `${result + (result === '' ? '' : ' and ') + object[key].field} ${key} ${object[key].value}`;
    } else {
      result = result + (result === '' ? '' : ' and ') + createSentance(object[key], runOnSentence);
    }
  }
  return result;
};

const determineIcon = (resultValue) => {
  // "icon" :
  // "fa fa-ban"                    <- grey crossed circle
  // "pficon pficon-ok"             <- green check circle
  // "pficon pficon-error-circle-o" <- red x circle
  switch (resultValue) {
    case 'deny':
      return 'pficon pficon-error-circle-o';
    case 'N/A':
      return 'fa fa-ban';
    case 'allow':
      return 'pficon pficon-ok';
    case true: // this happens in the scope / expression level
      return 'pficon pficon-ok'; // "fa fa-bell";
    case false:
      return 'pficon pficon-error-circle-o'; // "fa fa-bell-o";
    default:
      return 'fa fa-ban'; // "fa fa-bell-slash";
  }
};

const determineTextColor = (resultValue, text) => {
  switch (resultValue) {
    case true: // this happens in the scope / expression level
      return `\u003cfont color="green"\u003e${text}\u003c/font\u003e`;
    case false:
      return `\u003cfont color="red"\u003e${text}\u003c/font\u003e`;
    default:
      return text;
  }
};
