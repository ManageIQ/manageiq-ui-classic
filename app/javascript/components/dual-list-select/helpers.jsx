/* *********** HELPERS FOR FORMS *********** */

/* Get keys of options array
  Options:
    [
      {key: 'key', label: 'label'},
      {key: 'key1', label: 'label1'},
    ]
  returns ['key', 'key1']
*/
export const getKeys = value => (Array.isArray(value) ? value.map(({ key }) => key) : []);

/* Returns array of options excluding value options */
export const filterOptions = (options, value) => options.filter(({ key }) => !value.includes(key));

/* Returns items from newValue which are not in value */
export const filterValues = (newValue, value) => newValue.filter(item => !value.includes(item));
