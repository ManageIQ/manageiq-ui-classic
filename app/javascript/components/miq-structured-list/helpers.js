/* eslint-disable no-eval */
import { TreeViewRedux } from '../tree-view';
import CatalogResource from '../data-tables/catalog-resource';
import MiqDataTable from '../miq-data-table';

export const InputTypes = {
  TEXTAREA: 'text_area',
  CHECKBOX: 'checkbox',
  COMPONENT: 'component',
  DROPDOWN: 'dropdown',
  CODEMIRROR: 'code_mirror',
  MARKDOWN: 'markdown',
};

export const DynamicReactComponents = {
  TREE_VIEW_REDUX: TreeViewRedux,
  CATALOG_RESOURCE: CatalogResource,
  MIQ_DATA_TABLE: MiqDataTable,
};

const dataType = (data) => (data ? data.constructor.name.toString() : undefined);

export const isObject = (item) => dataType(item) === 'Object';

export const isArray = (item) => dataType(item) === 'Array';

export const isSubItem = (item) => item.sub_items && item.sub_items.length > 0;

/** When the props named mode includes 'miq_summary', then then summary contains row/cell click events. */
export const hasClickEvents = (mode) => mode.includes('miq_summary');

export const hasInput = (item) => item && Object.keys(item).includes('input');

export const rowClickEvent = (rows, index) => {
  const item = rows[index];
  if (item && item.onclick) {
    miqSparkleOn();
    eval(item.onclick);
    miqSparkleOff();
  }
};
