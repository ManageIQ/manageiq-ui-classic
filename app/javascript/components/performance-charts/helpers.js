export const getYAxisValue = (format, value) => {
  // eslint-disable-next-line no-useless-escape
  if (format) {
    const tmp = /^([0-9\,\.]+)(.*)/.exec(ManageIQ.charts.formatters[format.function].c3(format.options)(value));
    return [`${numeral(tmp[1]).value()}${tmp[2]}`];
  }
  return value;
};
