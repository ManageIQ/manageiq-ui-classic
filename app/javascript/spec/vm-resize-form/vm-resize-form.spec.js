import React from 'react';
import toJson from 'enzyme-to-json';
import fetchMock from 'fetch-mock';
import { shallow } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { mount } from '../helpers/mountForm';
import VmResizeForm from '../../components/vm-resize-form/vm-resize-form';

require('../helpers/miqSparkle.js');

describe('vm resize form component', () => {
  const response = {
    data: {
      form_schema: {
        fields: [
          {
            component: 'text-field',
            name: 'memory',
            id: 'memory',
            label: _('Memory (GiB)'),
            initialValue: 2,
            isRequired: true,
            validate: [{ type: 'required' }],
          },
        ],
      },
    },
  };

  beforeEach(() => {
    fetchMock.once('/api/vms/1?option_action=resize', response);
  });

  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  it('should render form', () => {
    const wrapper = shallow(<VmResizeForm />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render a resize form', async(done) => {
    let wrapper;
    await act(async() => {
      wrapper = mount(<VmResizeForm recordId="1" vmCloudResizeFormId="53" />);
    });

    setImmediate(() => {
      wrapper.update();
      expect(toJson(wrapper)).toMatchSnapshot();
      done();
    });
  });

  it('should submit resize API call', async(done) => {
    const submitData = {
      action: 'resize',
      resource: {
        resizeValues: { flavor: 'bh1-16x1600' },
      },
      resizeFormId: '53',
    };
    fetchMock.postOnce('/api/vms/1', submitData);
    let wrapper;
    await act(async() => {
      wrapper = mount(<VmResizeForm recordId="1" vmCloudResizeFormId="53" />);
    });
    expect(toJson(wrapper)).toMatchSnapshot();
    done();
  });
});
