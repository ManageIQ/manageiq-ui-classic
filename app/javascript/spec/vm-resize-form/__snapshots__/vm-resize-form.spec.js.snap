// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vm resize form component should render a resize form 1`] = `
<Provider
  store={
    Object {
      "asyncReducers": Object {
        "FormButtons": [Function],
        "miqCustomTabReducer": [Function],
        "notificationReducer": [Function],
      },
      "dispatch": [Function],
      "getState": [Function],
      "injectReducers": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <VmResizeForm
    recordId="1"
    vmCloudResizeFormId="53"
  >
    <Grid>
      <FlexGrid>
        <GridSettings
          mode="flexbox"
          subgrid={false}
        >
          <div
            className="bx--grid"
          >
            <Connect(MiqFormRenderer)
              onCancel={[Function]}
              onSubmit={[Function]}
              schema={
                Object {
                  "fields": Array [
                    Array [
                      Object {
                        "component": "text-field",
                        "id": "memory",
                        "initialValue": 2,
                        "isRequired": true,
                        "label": "Memory (GiB)",
                        "name": "memory",
                        "validate": Array [
                          Object {
                            "type": "required",
                          },
                        ],
                      },
                    ],
                  ],
                }
              }
            >
              <MiqFormRenderer
                buttonsLabels={Object {}}
                canReset={false}
                className="form-react"
                componentMapper={
                  Object {
                    "checkbox": [Function],
                    "code-editor": [Function],
                    "date-picker": [Function],
                    "dual-list-select": [Function],
                    "edit-password-field": [Function],
                    "field-array": [Function],
                    "file-upload": [Function],
                    "font-icon-picker": [Function],
                    "font-icon-picker-ddf": [Function],
                    "multi-select": [Function],
                    "password-field": [Function],
                    "plain-text": [Function],
                    "radio": [Function],
                    "select": [Function],
                    "slider": [Function],
                    "sub-form": [Function],
                    "switch": [Function],
                    "tabs": [Function],
                    "text-field": [Function],
                    "textarea": [Function],
                    "time-picker": [Function],
                    "tree-selector": [Function],
                    "tree-view": [Function],
                    "validate-credentials": [Function],
                    "wizard": [Function],
                  }
                }
                disableSubmit={
                  Array [
                    "pristine",
                    "invalid",
                  ]
                }
                dispatch={[Function]}
                onCancel={[Function]}
                onSubmit={[Function]}
                schema={
                  Object {
                    "fields": Array [
                      Array [
                        Object {
                          "component": "text-field",
                          "id": "memory",
                          "initialValue": 2,
                          "isRequired": true,
                          "label": "Memory (GiB)",
                          "name": "memory",
                          "validate": Array [
                            Object {
                              "type": "required",
                            },
                          ],
                        },
                      ],
                    ],
                  }
                }
                showFormControls={true}
              >
                <FormRenderer
                  FormTemplate={[Function]}
                  clearOnUnmount={false}
                  componentMapper={
                    Object {
                      "checkbox": [Function],
                      "code-editor": [Function],
                      "date-picker": [Function],
                      "dual-list-select": [Function],
                      "edit-password-field": [Function],
                      "field-array": [Function],
                      "file-upload": [Function],
                      "font-icon-picker": [Function],
                      "font-icon-picker-ddf": [Function],
                      "multi-select": [Function],
                      "password-field": [Function],
                      "plain-text": [Function],
                      "radio": [Function],
                      "select": [Function],
                      "slider": [Function],
                      "spy-field": [Function],
                      "sub-form": [Function],
                      "switch": [Function],
                      "tabs": [Function],
                      "text-field": [Function],
                      "textarea": [Function],
                      "time-picker": [Function],
                      "tree-selector": [Function],
                      "tree-view": [Function],
                      "validate-credentials": [Function],
                      "wizard": [Function],
                    }
                  }
                  dispatch={[Function]}
                  initialValues={Object {}}
                  onCancel={[Function]}
                  onReset={[Function]}
                  onSubmit={[Function]}
                  schema={
                    Object {
                      "fields": Array [
                        Array [
                          Object {
                            "component": "text-field",
                            "id": "memory",
                            "initialValue": 2,
                            "isRequired": true,
                            "label": "Memory (GiB)",
                            "name": "memory",
                            "validate": Array [
                              Object {
                                "type": "required",
                              },
                            ],
                          },
                        ],
                        Object {
                          "component": "spy-field",
                          "initialize": undefined,
                          "name": "spy-field",
                        },
                      ],
                    }
                  }
                >
                  <ReactFinalForm
                    decorators={
                      Array [
                        [Function],
                      ]
                    }
                    dispatch={[Function]}
                    initialValues={Object {}}
                    mutators={
                      Object {
                        "concat": [Function],
                        "insert": [Function],
                        "move": [Function],
                        "pop": [Function],
                        "push": [Function],
                        "remove": [Function],
                        "removeBatch": [Function],
                        "shift": [Function],
                        "swap": [Function],
                        "unshift": [Function],
                        "update": [Function],
                      }
                    }
                    onSubmit={[Function]}
                    render={[Function]}
                    subscription={
                      Object {
                        "pristine": true,
                        "submitting": true,
                        "valid": true,
                      }
                    }
                  >
                    <Component
                      formFields={
                        Array [
                          Array [
                            <SingleField
                              component="text-field"
                              id="memory"
                              initialValue={2}
                              isRequired={true}
                              label="Memory (GiB)"
                              name="memory"
                              validate={
                                Array [
                                  Object {
                                    "type": "required",
                                  },
                                ]
                              }
                            />,
                          ],
                          <SingleField
                            component="spy-field"
                            name="spy-field"
                          />,
                        ]
                      }
                      schema={
                        Object {
                          "fields": Array [
                            Array [
                              Object {
                                "component": "text-field",
                                "id": "memory",
                                "initialValue": 2,
                                "isRequired": true,
                                "label": "Memory (GiB)",
                                "name": "memory",
                                "validate": Array [
                                  Object {
                                    "type": "required",
                                  },
                                ],
                              },
                            ],
                            Object {
                              "component": "spy-field",
                              "initialize": undefined,
                              "name": "spy-field",
                            },
                          ],
                        }
                      }
                    >
                      <WrappedFormTemplate
                        canReset={false}
                        cancelLabel="Cancel"
                        disableSubmit={
                          Array [
                            "pristine",
                            "invalid",
                          ]
                        }
                        formFields={
                          Array [
                            Array [
                              <SingleField
                                component="text-field"
                                id="memory"
                                initialValue={2}
                                isRequired={true}
                                label="Memory (GiB)"
                                name="memory"
                                validate={
                                  Array [
                                    Object {
                                      "type": "required",
                                    },
                                  ]
                                }
                              />,
                            ],
                            <SingleField
                              component="spy-field"
                              name="spy-field"
                            />,
                          ]
                        }
                        formWrapperProps={
                          Object {
                            "className": "form-react",
                          }
                        }
                        resetLabel="Reset"
                        schema={
                          Object {
                            "fields": Array [
                              Array [
                                Object {
                                  "component": "text-field",
                                  "id": "memory",
                                  "initialValue": 2,
                                  "isRequired": true,
                                  "label": "Memory (GiB)",
                                  "name": "memory",
                                  "validate": Array [
                                    Object {
                                      "type": "required",
                                    },
                                  ],
                                },
                              ],
                              Object {
                                "component": "spy-field",
                                "initialize": undefined,
                                "name": "spy-field",
                              },
                            ],
                          }
                        }
                        showFormControls={true}
                        submitLabel="Save"
                      >
                        <FormTemplate
                          Button={[Function]}
                          ButtonGroup={[Function]}
                          Description={[Function]}
                          FormWrapper={[Function]}
                          Header={[Function]}
                          Title={[Function]}
                          buttonOrder={
                            Array [
                              "submit",
                              "reset",
                              "cancel",
                            ]
                          }
                          canReset={false}
                          cancelLabel="Cancel"
                          disableSubmit={
                            Array [
                              "pristine",
                              "invalid",
                            ]
                          }
                          formFields={
                            Array [
                              Array [
                                <SingleField
                                  component="text-field"
                                  id="memory"
                                  initialValue={2}
                                  isRequired={true}
                                  label="Memory (GiB)"
                                  name="memory"
                                  validate={
                                    Array [
                                      Object {
                                        "type": "required",
                                      },
                                    ]
                                  }
                                />,
                              ],
                              <SingleField
                                component="spy-field"
                                name="spy-field"
                              />,
                            ]
                          }
                          formWrapperProps={
                            Object {
                              "className": "form-react",
                            }
                          }
                          resetLabel="Reset"
                          schema={
                            Object {
                              "fields": Array [
                                Array [
                                  Object {
                                    "component": "text-field",
                                    "id": "memory",
                                    "initialValue": 2,
                                    "isRequired": true,
                                    "label": "Memory (GiB)",
                                    "name": "memory",
                                    "validate": Array [
                                      Object {
                                        "type": "required",
                                      },
                                    ],
                                  },
                                ],
                                Object {
                                  "component": "spy-field",
                                  "initialize": undefined,
                                  "name": "spy-field",
                                },
                              ],
                            }
                          }
                          showFormControls={true}
                          submitLabel="Save"
                        >
                          <Form
                            className="form-react"
                            onSubmit={[Function]}
                          >
                            <Form
                              className="form-0-2-3 form-react"
                              noValidate={true}
                              onSubmit={[Function]}
                            >
                              <form
                                className="bx--form form-0-2-3 form-react"
                                noValidate={true}
                                onSubmit={[Function]}
                              >
                                 
                                <SingleField
                                  component="text-field"
                                  id="memory"
                                  initialValue={2}
                                  isRequired={true}
                                  key="memory"
                                  label="Memory (GiB)"
                                  name="memory"
                                  validate={
                                    Array [
                                      Object {
                                        "type": "required",
                                      },
                                    ]
                                  }
                                >
                                  <FormConditionWrapper
                                    field={
                                      Object {
                                        "component": "text-field",
                                        "id": "memory",
                                        "initialValue": 2,
                                        "isRequired": true,
                                        "label": "Memory (GiB)",
                                        "name": "memory",
                                        "validate": Array [
                                          Object {
                                            "type": "required",
                                          },
                                        ],
                                      }
                                    }
                                  >
                                    <FormFieldHideWrapper
                                      hideField={false}
                                    >
                                      <TextField
                                        component="text-field"
                                        id="memory"
                                        initialValue={2}
                                        isRequired={true}
                                        label="Memory (GiB)"
                                        name="memory"
                                        validate={
                                          Array [
                                            Object {
                                              "type": "required",
                                            },
                                          ]
                                        }
                                      >
                                        <TextInput
                                          id="memory"
                                          invalid={false}
                                          invalidText=""
                                          key="memory"
                                          labelText={
                                            <IsRequired>
                                              Memory (GiB)
                                            </IsRequired>
                                          }
                                          name="memory"
                                          onBlur={[Function]}
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          value={2}
                                          warn={false}
                                          warnText=""
                                        >
                                          <div
                                            className="bx--form-item bx--text-input-wrapper"
                                          >
                                            <label
                                              className="bx--label"
                                              htmlFor="memory"
                                            >
                                              <IsRequired>
                                                <span
                                                  aria-hidden="true"
                                                  className="ddorg__carbon-component-mapper_is-required isRequired-0-2-4"
                                                >
                                                  *
                                                </span>
                                                Memory (GiB)
                                              </IsRequired>
                                            </label>
                                            <div
                                              className="bx--text-input__field-outer-wrapper"
                                            >
                                              <div
                                                className="bx--text-input__field-wrapper"
                                                data-invalid={null}
                                              >
                                                <input
                                                  className="bx--text-input bx--text-input--md"
                                                  disabled={false}
                                                  id="memory"
                                                  name="memory"
                                                  onBlur={[Function]}
                                                  onChange={[Function]}
                                                  onClick={[Function]}
                                                  onFocus={[Function]}
                                                  type="text"
                                                  value={2}
                                                />
                                              </div>
                                            </div>
                                          </div>
                                        </TextInput>
                                      </TextField>
                                    </FormFieldHideWrapper>
                                  </FormConditionWrapper>
                                </SingleField>
                                <SingleField
                                  component="spy-field"
                                  key="spy-field"
                                  name="spy-field"
                                >
                                  <FormConditionWrapper
                                    field={
                                      Object {
                                        "component": "spy-field",
                                        "initialize": undefined,
                                        "name": "spy-field",
                                      }
                                    }
                                  >
                                    <FormFieldHideWrapper
                                      hideField={false}
                                    >
                                      <SpyField
                                        component="spy-field"
                                        name="spy-field"
                                      >
                                        <FormSpy
                                          onChange={[Function]}
                                          subscription={
                                            Object {
                                              "pristine": true,
                                              "valid": true,
                                            }
                                          }
                                        />
                                      </SpyField>
                                    </FormFieldHideWrapper>
                                  </FormConditionWrapper>
                                </SingleField>
                                <FormSpy>
                                  <FormControls
                                    Button={[Function]}
                                    ButtonGroup={[Function]}
                                    FormSpy={[Function]}
                                    buttonOrder={
                                      Array [
                                        "submit",
                                        "reset",
                                        "cancel",
                                      ]
                                    }
                                    canReset={false}
                                    cancelLabel="Cancel"
                                    disableSubmit={true}
                                    formFields={
                                      Array [
                                        Array [
                                          <SingleField
                                            component="text-field"
                                            id="memory"
                                            initialValue={2}
                                            isRequired={true}
                                            label="Memory (GiB)"
                                            name="memory"
                                            validate={
                                              Array [
                                                Object {
                                                  "type": "required",
                                                },
                                              ]
                                            }
                                          />,
                                        ],
                                        <SingleField
                                          component="spy-field"
                                          name="spy-field"
                                        />,
                                      ]
                                    }
                                    formSpyProps={
                                      Object {
                                        "active": undefined,
                                        "dirty": false,
                                        "dirtyFields": Object {},
                                        "dirtyFieldsSinceLastSubmit": Object {
                                          "memory": true,
                                        },
                                        "dirtySinceLastSubmit": false,
                                        "error": undefined,
                                        "errors": Object {},
                                        "form": Object {
                                          "batch": [Function],
                                          "blur": [Function],
                                          "change": [Function],
                                          "destroyOnUnregister": false,
                                          "focus": [Function],
                                          "getFieldState": [Function],
                                          "getRegisteredFields": [Function],
                                          "getState": [Function],
                                          "initialize": [Function],
                                          "isValidationPaused": [Function],
                                          "mutators": Object {
                                            "concat": [Function],
                                            "insert": [Function],
                                            "move": [Function],
                                            "pop": [Function],
                                            "push": [Function],
                                            "remove": [Function],
                                            "removeBatch": [Function],
                                            "shift": [Function],
                                            "swap": [Function],
                                            "unshift": [Function],
                                            "update": [Function],
                                          },
                                          "pauseValidation": [Function],
                                          "registerField": [Function],
                                          "reset": [Function],
                                          "resetFieldState": [Function],
                                          "restart": [Function],
                                          "resumeValidation": [Function],
                                          "setConfig": [Function],
                                          "submit": [Function],
                                          "subscribe": [Function],
                                        },
                                        "hasSubmitErrors": false,
                                        "hasValidationErrors": false,
                                        "initialValues": Object {
                                          "memory": 2,
                                        },
                                        "invalid": false,
                                        "modified": Object {
                                          "memory": false,
                                        },
                                        "modifiedSinceLastSubmit": false,
                                        "pristine": true,
                                        "submitError": undefined,
                                        "submitErrors": undefined,
                                        "submitFailed": false,
                                        "submitSucceeded": false,
                                        "submitting": false,
                                        "touched": Object {
                                          "memory": false,
                                        },
                                        "valid": true,
                                        "validating": false,
                                        "values": Object {
                                          "memory": 2,
                                        },
                                        "visited": Object {
                                          "memory": false,
                                        },
                                      }
                                    }
                                    onCancel={[Function]}
                                    onReset={[Function]}
                                    resetLabel="Reset"
                                    schema={
                                      Object {
                                        "fields": Array [
                                          Array [
                                            Object {
                                              "component": "text-field",
                                              "id": "memory",
                                              "initialValue": 2,
                                              "isRequired": true,
                                              "label": "Memory (GiB)",
                                              "name": "memory",
                                              "validate": Array [
                                                Object {
                                                  "type": "required",
                                                },
                                              ],
                                            },
                                          ],
                                          Object {
                                            "component": "spy-field",
                                            "initialize": undefined,
                                            "name": "spy-field",
                                          },
                                        ],
                                      }
                                    }
                                    submitLabel="Save"
                                  >
                                    <ButtonGroup>
                                      <ButtonSet
                                        className=""
                                      >
                                        <div
                                          className="bx--btn-set"
                                        >
                                          <Button
                                            buttonType="submit"
                                            disabled={true}
                                            key="form-submit"
                                            label="Save"
                                            type="submit"
                                            variant="primary"
                                          >
                                            <Button
                                              disabled={true}
                                              kind="primary"
                                              type="submit"
                                              variant="primary"
                                            >
                                              <button
                                                aria-describedby={null}
                                                aria-pressed={null}
                                                className="bx--btn bx--btn--primary bx--btn--disabled"
                                                disabled={true}
                                                onBlur={[Function]}
                                                onClick={[Function]}
                                                onFocus={[Function]}
                                                onMouseEnter={[Function]}
                                                onMouseLeave={[Function]}
                                                tabIndex={0}
                                                type="submit"
                                                variant="primary"
                                              >
                                                Save
                                              </button>
                                            </Button>
                                          </Button>
                                          <Button
                                            buttonType="cancel"
                                            key="form-cancel"
                                            label="Cancel"
                                            onClick={[Function]}
                                            type="button"
                                          >
                                            <Button
                                              kind="secondary"
                                              onClick={[Function]}
                                              type="button"
                                            >
                                              <button
                                                aria-describedby={null}
                                                aria-pressed={null}
                                                className="bx--btn bx--btn--secondary"
                                                disabled={false}
                                                onBlur={[Function]}
                                                onClick={[Function]}
                                                onFocus={[Function]}
                                                onMouseEnter={[Function]}
                                                onMouseLeave={[Function]}
                                                tabIndex={0}
                                                type="button"
                                              >
                                                Cancel
                                              </button>
                                            </Button>
                                          </Button>
                                        </div>
                                      </ButtonSet>
                                    </ButtonGroup>
                                  </FormControls>
                                </FormSpy>
                                 
                              </form>
                            </Form>
                          </Form>
                        </FormTemplate>
                      </WrappedFormTemplate>
                    </Component>
                  </ReactFinalForm>
                </FormRenderer>
              </MiqFormRenderer>
            </Connect(MiqFormRenderer)>
          </div>
        </GridSettings>
      </FlexGrid>
    </Grid>
  </VmResizeForm>
</Provider>
`;

exports[`vm resize form component should render form 1`] = `""`;

exports[`vm resize form component should submit resize API call 1`] = `
<Provider
  store={
    Object {
      "asyncReducers": Object {
        "FormButtons": [Function],
        "miqCustomTabReducer": [Function],
        "notificationReducer": [Function],
      },
      "dispatch": [Function],
      "getState": [Function],
      "injectReducers": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <VmResizeForm
    recordId="1"
    vmCloudResizeFormId="53"
  />
</Provider>
`;
